contributors/design-proposals/controller-ref.md

```diff
@@ -1,100 +1,304 @@
-For debugging purposes we want to add an `adoptionTime` annotation prefixed with `kubernetes.io/` which will keep the time of last controller ownership transfer.
-
-# Upgrade/downgrade procedure
-
-Because `ControllerRef` will be a part of `OwnerReference` effort it will have the same upgrade/downgrade procedures.
-
-# Orphaning/adoption
-
-Because `ControllerRef` will be a part of `OwnerReference` effort it will have the same orphaning/adoption procedures.
-
-Controllers will orphan objects they own in two cases:
-* Change of label/selector causing selector to stop matching labels (executed by the controller)
-* Deletion of a controller with `Orphaning=true` (executed by the GarbageCollector)
-
-We will need a secondary orphaning mechanism in case of unclean controller deletion:
-* GarbageCollector will remove `ControllerRef` from objects that no longer points to existing controllers
-
-Controller will adopt (set `Controller` field in the `OwnerReference` that points to it) an object whose labels match its selector iff:
-* there are no `OwnerReferences` with `Controller` set to true in `OwnerReferences` array
-* `DeletionTimestamp` is not set
-and
-* Controller is the first controller that will manage to adopt the Pod from all Controllers that have matching label selector and don't have `DeletionTimestamp` set.
-
-By design there are possible races during adoption if multiple controllers can own a given object.
-
-To prevent re-adoption of an object during deletion the `DeletionTimestamp` will be set when deletion is starting. When a controller has a non-nil `DeletionTimestamp` it won't take any actions except updating its `Status` (in particular it won't adopt any objects).
-
-# Implementation plan (sketch):
-
-* Add API field for `Controller`,
-* Extend `OwnerReference` adoption procedure to set a `Controller` field in one of the owners,
-* Update all affected controllers to respect `ControllerRef`.
-
-Necessary related work:
-* `OwnerReferences` are correctly added/deleted,
-* GarbageCollector removes dangling references,
-* Controllers don't take any meaningful actions when `DeletionTimestamps` is set.
-
-# Considered alternatives
+A ControllerRef is thus defined as an OwnerReference with `Controller=true`.
+Each object may have at most one ControllerRef in its list of OwnerReferences.
+The validator for OwnerReferences lists will fail any update that would violate
+this invariant.
+
+# Behavior
+
+This section summarizes the intended behavior for existing controllers.
+It can also serve as a guide for respecting ControllerRef when writing new
+controllers.
+
+## The Three Laws of Controllers
+
+All controllers that manage collections of objects should obey the following
+rules.
+
+1. **Take ownership**
+
+   A controller must claim *ownership* of any objects it creates by adding a
+   ControllerRef, and may also claim ownership of an object it didn't create,
+   as long as the object has no existing ControllerRef (i.e. it is an *orphan*).
```

Comment:
I general, I worry it will get confusing if I add conditions to the definition of "orphan". If an object was created before ControllerRef, I would still call it an orphan since it meets the simple definition in terms of having a ControllerRef. If you think we need to talk more about the transition period, I think it should go in the Upgrading section.

Specifically, this list of rules is intended to be applicable in the long-term in a post-ControllerRef world where all controllers respect it. These rules explain how you should behave if you're adding a new controller to that world.